### ISSUE TITLE
GCP Terraform  dataproc - ignores internal_ip_only value as False

### ISSUE THREAD

**Issue by tenstriker** [hashicorp/terraform-provider-google#17436]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to a user, that user is claiming responsibility for the issue.
* Customers working with a Google Technical Account Manager or Customer Engineer can ask them to [reach out internally](https://github.com/hashicorp/terraform-provider-google/wiki/Customer-Contact#raising-gcp-internal-issues-with-the-provider-development-team) to expedite investigation and resolution of this issue.

### Terraform Version

google provider 4.75.0 and 5.17.0
Terraform v1.7.4

### Affected Resource(s)

google_dataproc_cluster

### Terraform Configuration

Happens with unchanged plan. Just trying to re-execute things. Only thing that happens on re-applying same stuff is update of terraform iteself 

```tf apply

```

### Debug Output

Just pasting snippet as debug out has lot of confidential info. Issue is around not respecting default or explicit value of `internal_ip_only` when it sets to False . (it is set to false by default) instead upon Terraform Apply it consider it to be true. (based on error message)

2024-02-27T02:22:26.291Z [DEBUG] provider.terraform-provider-google_v5.17.0_x5: 2024/02/27 02:22:26 [DEBUG] parent context canceled, cleaning up batcher batches
2024-02-27T02:22:26.291Z [DEBUG] provider.terraform-provider-google_v5.17.0_x5: 2024/02/27 02:22:26 [DEBUG] Stopping batcher "Service Usage"
2024-02-27T02:22:26.333Z [WARN]  Provider "registry.terraform.io/hashicorp/google" produced an invalid plan for google_dataproc_cluster.datascience-cluster, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .graceful_decommission_timeout: planned value cty.StringVal("0s") for a non-computed attribute
      - .virtual_cluster_config: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].auxiliary_node_groups: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].dataproc_metric_config: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].master_config[0].disk_config: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].worker_config[0].disk_config: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].gce_cluster_config[0].service_account_scopes: planned value cty.SetVal([]cty.Value{cty.StringVal("[https://www.googleapis.com/auth/cloud-platform"),](https://www.googleapis.com/auth/cloud-platform%22),) cty.StringVal("[https://www.googleapis.com/auth/cloud.useraccounts.readonly"),](https://www.googleapis.com/auth/cloud.useraccounts.readonly%22),) cty.StringVal("[https://www.googleapis.com/auth/devstorage.read_write"),](https://www.googleapis.com/auth/devstorage.read_write%22),) cty.StringVal("[https://www.googleapis.com/auth/logging.write")}](https://www.googleapis.com/auth/logging.write%22)%7D)) does not match config value cty.SetVal([]cty.Value{cty.StringVal("cloud-platform"), cty.StringVal("logging-write"), cty.StringVal("storage-rw"), cty.StringVal("useraccounts-ro")})
      - .cluster_config[0].gce_cluster_config[0].**internal_ip_only**: planned value cty.False for a non-computed attribute
      - .cluster_config[0].gce_cluster_config[0].node_group_affinity: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].gce_cluster_config[0].reservation_affinity: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].gce_cluster_config[0].shielded_instance_config: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
      - .cluster_config[0].preemptible_worker_config: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
2024-02-27T02:22:26.333Z [INFO]  Starting apply for google_dataproc_cluster.datascience-cluster

### Steps to reproduce

1. `terraform apply`

**Comment by zli82016** [comment]
@tenstriker , can you please provide the configuration to reproduce the issue?

After upgrading google provider from  4.75.0 to 5.17.0 and then running the command `terraform apply`, an error occurs?

**Comment by tenstriker** [comment]
> @tenstriker , can you please provide the configuration to reproduce the issue?
> 
> After upgrading google provider from 4.75.0 to 5.17.0 and then running the command `terraform apply`, an error occurs. Is that the issue?

No it's actually happens with both. It used to work with 4.75.0 but suddenly started failing so I tried latest version and it fails again now. I wonder just the gcp backend api changed. means TF was never sending the value but in backend it was considered False by default but recently considering it to be True now? Debug log is taken with version 5.17 which clearly doesn't send this param value in request body.

**Comment by zli82016** [comment]
Thanks for the information, @tenstriker . Is it possible to provide the configuration?

**Comment by tenstriker** [comment]
sure, I added in OP. I tried tweaking some parameters of cluster_config like removing autoscaling, optional_components, changing image version etc to no avail

**Comment by zli82016** [comment]
Forward this issue to the service team to check the reason for the error message.

**Comment by tenstriker** [comment]
thanks. fyi, I ran both versions (4.75.0 and 5.17.0) with Terraform v1.7.4 and none of them seem to have `internal_ip_only` in request body.

**Comment by tenstriker** [comment]
I think this might also have to do with image version I am using `2.2.0-RC3-debian11`

**Comment by tenstriker** [comment]
Tested another 2.2.* image 2.2.3-debian12 . so seem like dataproc 2.2.* image versions may have to do with this?

**Comment by tenstriker** [comment]
I added some more updates in OP

**Comment by cnauroth** [comment]
I don't think this is a bug in Terraform, and perhaps we can close this issue. Instead, I think this was a change introduced in Dataproc for stronger security defaults:

https://cloud.google.com/dataproc/docs/release-notes#February_16_2024

> Dataproc on Compute Engine: The [internalIpOnly](https://cloud.google.com/dataproc/docs/reference/rest/v1/ClusterConfig#GceClusterConfig.FIELDS.internal_ip_only) cluster configuration setting now defaults to true for clusters created with [2.2 image versions](https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-release-2.2). Also see [Create a Dataproc cluster with internal IP addresses only](https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/network#create_a_cluster_with_internal_ip_addresses_only).

The timing of that release roughly correlates with the date this issue was created.

**Comment by cnauroth** [comment]
...although #18503 suggests that even if you explicitly set `internal_ip_only = false`, it's not respected. That part would definitely be a bug.

**Comment by zli82016** [comment]
> ...although #18503 suggests that even if you explicitly set `internal_ip_only = false`, it's not respected. That part would definitely be a bug.

This is still a bug for this case and needs to be fixed.
